psydac/api/ast/fem.py

```diff
@@ -284,52 +251,21 @@ def __init__(self, expr, terminal_expr, spaces, tag=None, **kwargs):
             atomic_expr_field = {f:[] for f in fields}
             for i_row in range(0, n_rows):
                 for i_col in range(0, n_cols):
-                    d            = get_max_logical_partial_derivatives(terminal_expr[i_row,i_col])
-                    nderiv       = max(nderiv, max(d.values()))
-                    atoms        = _atomic(terminal_expr[i_row, i_col], cls=atoms_types+_logical_partial_derivatives)
-                    fields_atoms = [atom for atom in atoms if is_scalar_field(atom) or is_vector_field(atom)]
```

Comment:
can you keep this part, we need it, it's a bug that must be fixed not deleted

Reply:
I have only removed one line, where `fields_atoms` is defined.

Reply:
yes we need it, to know what atoms are used

Reply:
But the functions `is_scalar_field` and `is_vector_field` do not exist anymore!

Reply:
they need to be changed accordingly, and check if the test function in each atom are part of the fields, something like
`fields_atoms = [a for a in atoms if  get_test_function(a) in fields]`

Reply:
The fact that the code was wrong does not change the fact that it is dead code. It is **never** used in all our tests...

If you add some failing tests I will put this code back :wink:

Reply:
yes ,it should be fixed not deleted

Reply:
This should be fixed now.